### PR TITLE
Support custom changesets

### PR DESCRIPTION
**Pull request by zoedsoupe**
## Rationale

Support custom changesets with the `:changeset` option for the following functions:
- `create/2`
- `create!/2`
- `update/2`
- `update!/2`

Also add support for [`nix`](https://nixos.org) building and development environment. If you don't agree with this change, I can remove it from the repo.

Finally, `insert/2` and `insert!/2` behaviour should accept only built changesets or structs, to differenciate from the `create/2` and `create!/2` functions. What do you think about it?

See #15.

### CODE REVIEW

**Comment by joeljuca** [comment]
Zoey,

I’m not sure I understand the problem. Could you elaborate the problem of
using %__MODULE__{} to match the struct of a given schema?

**Comment by joeljuca** [comment]
@zoedsoupe I just saw that this is a PR, not an issue. Sorry – my previous comment came before I had a chance to review your PR, but I'll sure take some time to take a closer look at it tomorrow.

File: lib/swiss_schema.ex [hunk 1]
```diff
@@ -448,11 +467,13 @@ defmodule SwissSchema do
         all.(__MODULE__, opts)
       end
 
+      @default_changeset Function.capture(__MODULE__, :changeset, 2)
```
Comment (joeljuca) [inline lib/swiss_schema.ex:470]: Does it make `__MODULE__` a compile dependency of SwissSchema? Because if it does, I believe it becomes a cyclic dependency issue with unexpected effects.

I'm not sure about it being an issue, but if it is, we can just stick with `Function.capture/3` directly in each function, just like before.

File: lib/swiss_schema.ex [hunk 2]
```diff
@@ -533,25 +554,33 @@ defmodule SwissSchema do
       end
 
       @impl SwissSchema
-      def insert(%{} = params, opts \\ []) do
+      def insert(source, opts \\ [])
```
Comment (joeljuca) [inline lib/swiss_schema.ex:557]: This isn't ideal. Changing the API might cause bad effects in some code using this library in production.

If we're going to touch the API like that, we need to discuss and plan these changes and their releases accordingly.

For now, let's focus on supporting custom changesets in this PR.

File: test/swiss_schema_test.exs [hunk 3]
```diff
@@ -486,6 +528,66 @@ defmodule SwissSchemaTest do
...
+    end
+
+    test "accepts a custom Ecto changeset/2 function thru :changeset opt", %{user: user} do
+      user = User.update!(user, @update_params)
+
+      assert %User{} = user
+      assert is_integer(user.lucky_number)
+    end
+  end
+
```
Comment (joeljuca) [inline test/swiss_schema_test.exs:590]: Omg this is gold! 😍 But do you mind moving this to a dedicated PR? It would be nice to review these in a dedicated PR/discussion.

**Review by joeljuca — CHANGES_REQUESTED** [review]
Hi @zoedsoupe!

Thank you for your PR ❤️ I've left some comments and change requests for us to move forward. :)

File: lib/swiss_schema.ex [hunk 4]
```diff
@@ -533,25 +554,33 @@ defmodule SwissSchema do
...
+
+      def insert(%__MODULE__{} = schema, opts) do
```
[the remark this answers is not in the document]
Comment (zoedsoupe) [reply lib/swiss_schema.ex]: solved as we don’t have more compile-time dependencies on the source module/schema. i’ll create a issue for that.
